cdap-app-fabric/src/main/java/co/cask/cdap/config/PreferencesWrapper.java

```diff
@@ -0,0 +1,169 @@
+    return getPropertiesHelper(generateId(namespace, appId));
+  }
+
+  public Map<String, String> getProperties(String namespace, String appId, String programType, String programId) {
+    return getPropertiesHelper(generateId(namespace, appId, programType, programId));
+  }
+
+  public Map<String, String> getResolvedProperties() {
+    return getProperties();
+  }
+
+  public Map<String, String> getResolvedProperties(String namespace) {
+    Map<String, String> propMap = getResolvedProperties();
+    propMap.putAll(getProperties(namespace));
+    return propMap;
+  }
+
+  public Map<String, String> getResolvedProperties(String namespace, String appId) {
+    Map<String, String> propMap = getResolvedProperties(namespace);
+    propMap.putAll(getProperties(namespace, appId));
+    return propMap;
+  }
+
+  public Map<String, String> getResolvedProperties(String namespace, String appId, String programType,
+                                                   String programId) {
+    Map<String, String> propMap = getResolvedProperties(namespace, appId);
+    propMap.putAll(getProperties(namespace, appId, programType, programId));
+    return propMap;
+  }
+
+  public void setProperties(Map<String, String> propMap) {
+    setPropertiesHelper(generateId(), propMap);
+  }
+
+  public void setProperties(String namespace, Map<String, String> propMap) {
+    setPropertiesHelper(generateId(namespace), propMap);
+  }
+
+  public void setProperties(String namespace, String appId, Map<String, String> propMap) {
+    setPropertiesHelper(generateId(namespace, appId), propMap);
+  }
+
+  public void setProperties(String namespace, String appId, String programType, String programId,
+                            Map<String, String> propMap) {
+    setPropertiesHelper(generateId(namespace, appId, programType, programId), propMap);
+  }
+
+  public void deleteProperties() {
+    deletePropertiesHelper(generateId());
+  }
+
+  public void deleteProperties(String namespace) {
+    deletePropertiesHelper(generateId(namespace));
+  }
+
+  public void deleteProperties(String namespace, String appId) {
+    deletePropertiesHelper(generateId(namespace, appId));
+  }
+
+  public void deleteProperties(String namespace, String appId, String programType, String programId) {
+    deletePropertiesHelper(generateId(namespace, appId, programType, programId));
+  }
+
+  private String generateId() {
+    return getMultipartKey(INSTANCE_NAME);
+  }
+
+  private String generateId(String namespace) {
+    return getMultipartKey(INSTANCE_NAME, namespace);
+  }
+
+  private String generateId(String namespace, String appId) {
+    return getMultipartKey(INSTANCE_NAME, namespace, appId);
+  }
+
+  private String generateId(String namespace, String appId, String programType, String programId) {
+    return getMultipartKey(INSTANCE_NAME, namespace, appId, programType, programId);
+  }
+
+  private String getMultipartKey(String... parts) {
```

Comment:
There is some logic which is common. We don't have the VERSION byte here. Should we have one? 

If we want to extract out the common logic, where would be a good place to put it? Some final class with static method?